package java.lang;

// Exception class
public class Exception extends Throwable {
    static final long serialVersionUID = -3387516993124229948L;

    public Exception() {
        super();
    }

    public Exception(String message) {
        super(message);
    }

    public Exception(String message, Throwable cause) {
        super(message, cause);
    }

    public Exception(Throwable cause) {
        super(cause);
    }

    protected Exception(String message, Throwable cause, boolean enableSuppression, boolean     writableStackTrace) {
        super(message, cause, enableSuppression, writableStackTrace);
    }
}

// write your own exception class and extend Exception class
public class HandledException extends Exception {
    private String code;

    public HandledException(String code, String message) {
        super(message);
        this.setCode(code);
    }

    public HandledException(String code, String message, Throwable cause) {
        super(message, cause);
        this.setCode(code);
    }

    public String getCode() {
        return code;
    }

    public void setCode(String code) {
        this.code = code;
    }
}


public class MainClass {

    private static String DATABASE_EXCEPTION = "DATABASE_EXCEPTION";
    private static final Logger logger = Logger.getLogger(MainClass.class);

    public static void main(String[] args) {
        try {
            makeDatabaseConnection();
        } catch (HandledException e) {
            // Display custom message to the user
            System.out.println("Code: "+e.getCode()+” Exception Message : ”+e.getMessage());
            // Log the exception detail
            logger.error("Exception: ", e);
        }
    }

    static void makeDatabaseConnection() throws HandledException {
        String dbURL = "jdbc:sqlserver://localhost\sqlexpress";
        String userName = "sa";
        String password = "secret";
        Connection conn = null;
        try {
            conn = DriverManager.getConnection(dbURL, userName, password);
        } catch (SQLException e) {
            throw new HandledException(DATABASE_EXCEPTION,"Failed to connect to database", e);
        }
    }
}